### PR TITLE
Additional Nullability checks for deconstruction:

### PR DESCRIPTION
**Pull request by chsienki**
- Check 'this' param for extension deconstruct
- Re-infer the argument types for generic extension deconstruct
- Update return type with visited arguments
- Update tests

Closes https://github.com/dotnet/roslyn/issues/33006

### CODE REVIEW

**Comment by chsienki** [comment]
@dotnet/roslyn-compiler for review please

File: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs [hunk 1]
```diff
@@ -88591,9 +88594,50 @@ static void F(Pair<object?, object>? p)
     }
 }";
             var comp = CreateCompilation(source, options: WithNonNullTypesTrue());
-            // https://github.com/dotnet/roslyn/issues/33006: Check nullability of `this`
-            // for generic `Deconstruct` method and infer nullability of type arguments.
-            comp.VerifyDiagnostics();
+            comp.VerifyDiagnostics(
+                // (12,34): warning CS8604: Possible null reference argument for parameter 'p' in 'void E.Deconstruct<object, object>(Pair<object, object> p, out object t, out object u)'.
```
Comment (333fred) [inline src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs]: I understand why this is the warning we give, but this warning does not make much sense at first glance. I think it'd probably be worth introducing a custom message for extension deconstruction methods.

File: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs [hunk 2]
```diff
@@ -88612,17 +88656,73 @@ static class E
...
+        where T : class
+        => throw null!;
+
+    internal static void Deconstruct<T, U>(this Pair2<T, U> p, out T t, out U u)
+        where U : class
+        => throw null!;
+}
+class Program
+{
+    static void F(Pair<object?, Pair2<object, object?>?> p)
```
Comment (333fred) [inline src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs]: >Pair2 [](start = 32, length = 5)

Not sure I understand why you needed to introduce `Pair2`?

**Comment by 333fred** [comment]
Please add some tests for `Deconstruct` extension methods that accept a nullable `Pair`

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs [hunk 3]
```diff
@@ -5850,9 +5871,9 @@ public override BoundNode VisitAttribute(BoundAttribute node)
             return null;
         }
 
-        public override BoundNode VisitDeconstructValuePlaceholder(BoundDeconstructValuePlaceholder node)
+        public override BoundNode VisitExpressionWithNullability(BoundExpressionWithNullability node)
         {
-            SetNotNullResult(node);
+            ResultType = TypeWithState.Create(node.Type, node.NullableAnnotation == NullableAnnotation.Annotated ? NullableFlowState.MaybeNull : NullableFlowState.NotNull);
```
Comment (333fred) [inline src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs]: Note: this will need to be adjusted when features/nullable-api is merged (hopefully later today).

Comment (jcouv) [inline src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs]: Feels like there may be a bug here. Probably should be `TypeWithAnnotation.Create(node.Type, node.NullableAnnotation).ToTypeWithState()` instead. `ToTypeWithState` handles type parameters.
Please confirm and add a test to illustrate if applicable.

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs [hunk 4]
```diff
@@ -4509,11 +4509,31 @@ private void VisitDeconstructionArguments(ArrayBuilder<DeconstructionVariable> v
                     }
                     else
                     {
-                        // https://github.com/dotnet/roslyn/issues/33006: Check nullability of `this` argument.
-                    }
+                        // Check nullability for `this` parameter
```
Comment (jcouv) [inline src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs]: Is there a way to re-use the implementation of `VisitArguments` (line 2669) instead of duplicating some/most of it?

File: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs (hunk 2 above)
Comment (chsienki) [reply src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs]: Without it we return two identical warnings (same class, same parameter), which makes the test less obvious what's going on. I added Pair2 to clarify what's actually being reported.

---
In reply to: [276820962](https://github.com/dotnet/roslyn/pull/35016#discussion_r276820962) [](ancestors = 276820962)

File: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs (hunk 1 above)
Comment (chsienki) [reply src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs]: I switched it to be 8602. Turns out it made the code simpler, and aligned the extension method warnings with the instance method ones. Arguably we could still have a specific warning for these cases, but I'm not sure if it adds much.

---
In reply to: [276820010](https://github.com/dotnet/roslyn/pull/35016#discussion_r276820010) [](ancestors = 276820010)

**Comment by chsienki** [comment]
Added Deconstruction_ExtensionMethod_05

---
In reply to: [484677888](https://github.com/dotnet/roslyn/pull/35016#issuecomment-484677888) [](ancestors = 484677888)

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs (hunk 4 above)
Comment (chsienki) [reply src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs]: Not easily. Its a path I started down previously, but ended up being pretty messy too. Happy to give it another go though, if we think it'd be better factored.

---
In reply to: [277465231](https://github.com/dotnet/roslyn/pull/35016#discussion_r277465231) [](ancestors = 277465231)

**Comment by chsienki** [comment]
@dotnet/roslyn-compiler for a second round of review please

File: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs [hunk 5]
```diff
@@ -88837,7 +88891,7 @@ static class E
 }
 class Program
 {
-    static void F(Pair<object, object?>? p)
+    static void F(Pair<object, object?> p)
```
Comment (cston) [inline src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs]: >F(Pair<object, object?> p) [](start = 16, length = 26)

Please revert this change to ensure we're not reporting a warning for invocation of `Deconstruct`.

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs [hunk 6]
```diff
@@ -4808,23 +4808,33 @@ private void VisitDeconstructionArguments(ArrayBuilder<DeconstructionVariable> v
...
                 var deconstructMethod = invocation?.Method;
 
                 if ((object)deconstructMethod != null)
                 {
-                    if (!invocation.InvokedAsExtensionMethod)
-                    {
-                        _ = CheckPossibleNullReceiver(right);
-                    }
-                    else
+                    _ = CheckPossibleNullReceiver(right);
```
Comment (cston) [inline src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs]: >_ = CheckPossibleNullReceiver(right) [](start = 20, length = 36)

We should not report a warning when calling an extension method.

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs [hunk 7]
```diff
@@ -4808,23 +4808,33 @@ private void VisitDeconstructionArguments(ArrayBuilder<DeconstructionVariable> v
...
+                        // re-infer the deconstruct parameters based on the 'this' parameter 
+                        ArrayBuilder<BoundExpression> placeholderArgs = ArrayBuilder<BoundExpression>.GetInstance(n);
+                        placeholderArgs.Add(CreatePlaceholderIfNecessary(right, rightResultWithAnnotations));
+                        for (int i = 0; i < n; i++)
+                        {
+                            placeholderArgs.Add(new BoundExpressionWithNullability(variables[i].Expression.Syntax, variables[i].Expression, NullableAnnotation.Oblivious, conversion.DeconstructionInfo.OutputPlaceholders[i].Type));
+                        }
+                        deconstructMethod = InferMethodTypeArguments(invocation, deconstructMethod, placeholderArgs.ToImmutableAndFree());
 
-                    // https://github.com/dotnet/roslyn/issues/33006: Update `Deconstruct` method
```
Comment (cston) [inline src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs]: >Update `Deconstruct` method [](start = 70, length = 27)

Are we updating the `Deconstruct` method when it is an instance method?

For instance, in the following, we should not report a warning for `x.ToString()`.

```
class Pair<T, U>
{
    public void Deconstruct(out T t, out U u) => throw null!;
}

class Program
{
    static Pair<T, U> CreatePair<T, U>(T t, U u) => new Pair<T, U>();

    static void F(string? x, object? y)
    {
        if (x == null) return;
        var p = CreatePair(x, y);
        (x, y) = p;
        x.ToString(); // ok
        y.ToString(); // warning
    }
}
```

**Comment by cston** [comment]
>                             var parameter = parameters[i + offset];

Please move this ahead of the `if` and use in the `if` branch.

---
Refers to: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:4855 in b87e729. [](commit_id = b87e729ffc6c5c29eae98f3e393ca19f30087cb7, deletion_comment = False)

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs (hunk 7 above)
Comment (chsienki) [reply src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs]: Fixed and added Deconstruction_30 to cover this case

---
In reply to: [279097281](https://github.com/dotnet/roslyn/pull/35016#discussion_r279097281) [](ancestors = 279097281)

File: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs (hunk 1 above)
Comment (chsienki) [reply src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs]: Opened #35349 to track

---
In reply to: [278754133](https://github.com/dotnet/roslyn/pull/35016#discussion_r278754133) [](ancestors = 278754133,276820010)

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs [hunk 8]
```diff
@@ -4817,14 +4819,40 @@ private void VisitDeconstructionArguments(ArrayBuilder<DeconstructionVariable> v
...
+                        var parameter = deconstructMethod.Parameters[0];
+                        VisitArgumentConversion(
+                                right, conversion, parameter.RefKind, parameter, parameter.TypeWithAnnotations,
+                                new VisitArgumentResult(new VisitResult(rightResult, rightResultWithAnnotations), stateForLambda: default),
+                                extensionMethodThisArgument: true);
+
+                        if (deconstructMethod.IsGenericMethod)
+                        {
+                            // re-infer the deconstruct parameters based on the 'this' parameter 
+                            ArrayBuilder<BoundExpression> placeholderArgs = ArrayBuilder<BoundExpression>.GetInstance(n);
```
Comment (cston) [inline src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs]: >n [](start = 118, length = 1)

`n + 1`

File: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs [hunk 9]
```diff
@@ -88824,6 +88824,89 @@ static void F(object? x, object y)
...
+    static Pair<T, U> CreatePair<T, U>(T t, U u) => new Pair<T, U>();
+
+    static void F(string? x, object? y)
+    {
+        if (x == null) return;
+        var p = CreatePair(x, y);
+        (x, y) = p;
+        x.ToString(); // ok
+        y.ToString(); // warning
+    }
```
Comment (cston) [inline src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs]: >} [](start = 4, length = 1)

Are we testing re-inferring the `Deconstruct()` method for nested arguments? Perhaps something like:
```
    static void F(string? x, object? y)
    {
        if (x == null) return;
        var p = CreatePair(x, CreatePair(x, y));
        object? z;
        (z, (x, y)) = p;
        x.ToString(); // ok
        y.ToString(); // warning
    }
```

File: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs [hunk 10]
```diff
@@ -88899,9 +88985,83 @@ static void F(Pair<object?, object>? p)
...
+}
+class Program
+{
+    static void F(Pair<object?, Pair<object, object?>> p)
+    {
+        (object? x, (object y, object? z)) = p; 
+        x.ToString(); // 1
+        y.ToString(); 
+        z.ToString(); // 2
+    }
```
Comment (cston) [inline src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs]: >} [](start = 4, length = 1)

Are we testing that method type arguments are based on tracked state of `this` argument? For instance:
```
    static void F(string? x, object? y)
    {
        if (x == null) return;
        var p = CreatePair(x, CreatePair(x, y));
        object? z;
        (z, (x, y)) = p;
        x.ToString(); // ok
        y.ToString(); // warning
    }
```

File: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs (hunk 9 above)
Comment (chsienki) [reply src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs]: Added Deconstruction_31

---
In reply to: [279571543](https://github.com/dotnet/roslyn/pull/35016#discussion_r279571543) [](ancestors = 279571543)

File: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs (hunk 10 above)
Comment (chsienki) [reply src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs]: Added Deconstruction_ExtensionMethod_06

---
In reply to: [279572535](https://github.com/dotnet/roslyn/pull/35016#discussion_r279572535) [](ancestors = 279572535)
